compiler/shared/test/diff/NuMono/Differentiation.mls

```diff
@@ -0,0 +1,92 @@
+//│     fun getVal = concat("(",)(concat(concat(concat(((this).lhs).getVal,)(" + ",),)(((this).rhs).getVal,),)(")",),)
+//│   }
+//│   class Mul$5([lhs: Exp$1, rhs: Exp$1,]): Exp$1() {
+//│     fun derive = () => Sum$4(Mul$5(((this).lhs).derive(), (this).rhs,), Mul$5((this).lhs, ((this).rhs).derive(),),)
+//│     fun getVal = concat("(",)(concat(concat(concat(((this).lhs).getVal,)(" * ",),)(((this).rhs).getVal,),)(")",),)
+//│   }
+//│   Code(List((Sum$4(Variable$3("y",), Variable$3("x",),)).derive()))
+//│ }
+//│ Mono:
+//│ TypingUnit(NuTypeDef(class, Numeric$2, (), Tup(i: Var(Int)), (App(Var(Exp$1), Tup())), None, None, TypingUnit()), NuTypeDef(class, Sum$4, (), Tup(lhs: Var(Exp$1), rhs: Var(Exp$1)), (App(Var(Exp$1), Tup())), None, None, TypingUnit()), NuTypeDef(class, Mul$5, (), Tup(lhs: Var(Exp$1), rhs: Var(Exp$1)), (App(Var(Exp$1), Tup())), None, None, TypingUnit()), NuTypeDef(class, Variable$3, (), Tup(nm: Var(Str)), (App(Var(Exp$1), Tup())), None, None, TypingUnit()), NuTypeDef(class, Exp$1, (), Tup(), (), None, None, TypingUnit()), NuFunDef(None, derive$Sum$4, None, [], Lam(Tup(_: Var(this)), App(Var(Sum$4), Tup(_: Let(false, obj, Let(false, obj, Var(this), If(IfOpApp(Var(obj), Var(is), IfBlock(IfThen(App(Var(Sum$4), Tup(_: Var(lhs), _: Var(rhs))), Var(lhs)))), None)), If(IfOpApp(Var(obj), Var(is), IfBlock(IfThen(Var(Variable$3), App(Var(derive$Variable$3), Tup(_: Var(obj)))))), None)), _: Let(false, obj, Let(false, obj, Var(this), If(IfOpApp(Var(obj), Var(is), IfBlock(IfThen(App(Var(Sum$4), Tup(_: Var(lhs), _: Var(rhs))), Var(rhs)))), None)), If(IfOpApp(Var(obj), Var(is), IfBlock(IfThen(Var(Variable$3), App(Var(derive$Variable$3), Tup(_: Var(obj)))))), None)))))), NuFunDef(None, main$$5, None, [], Lam(Tup(), Let(false, obj, App(Var(Sum$4), Tup(_: App(Var(Variable$3), Tup(_: StrLit(y))), _: App(Var(Variable$3), Tup(_: StrLit(x))))), If(IfOpApp(Var(obj), Var(is), IfBlock(IfThen(Var(Sum$4), App(Var(derive$Sum$4), Tup(_: Var(obj)))))), None)))), NuFunDef(None, derive$Variable$3, None, [], Lam(Tup(_: Var(this)), App(Var(Numeric$2), Tup(_: IntLit(1))))), App(Var(main$$5), Tup()))
+//│ TypingUnit {
+//│   class Numeric$2([i: Int,]): Exp$1() {}
+//│   class Sum$4([lhs: Exp$1, rhs: Exp$1,]): Exp$1() {}
+//│   class Mul$5([lhs: Exp$1, rhs: Exp$1,]): Exp$1() {}
+//│   class Variable$3([nm: Str,]): Exp$1() {}
+//│   class Exp$1([]) {}
+//│   fun derive$Sum$4 = (this,) => Sum$4(let obj = let obj = this in if obj is ‹(Sum$4(lhs, rhs,)) then lhs› in if obj is ‹(Variable$3) then derive$Variable$3(obj,)›, let obj = let obj = this in if obj is ‹(Sum$4(lhs, rhs,)) then rhs› in if obj is ‹(Variable$3) then derive$Variable$3(obj,)›,)
+//│   fun main$$5 = () => let obj = Sum$4(Variable$3("y",), Variable$3("x",),) in if obj is ‹(Sum$4) then derive$Sum$4(obj,)›
+//│   fun derive$Variable$3 = (this,) => Numeric$2(1,)
+//│   Code(List(main$$5()))
+//│ }
+//│ ╔══[ERROR] Type mismatch in `case` expression:
```

Comment:
The generated pattern match of defunctionalization doesn't pass the type check because it doesn't match for all possible types of `Exp`. However in the full program no `Exp` flows into the pattern match other than `Variable`.
Should we:
\- Ignore the typing error?
\- Generate all possible cases?

Reply:
For now I think it would be better to generate a catch-all case that throws an "unreachable" exception.

Reply:
Added an `else error` branch. Is there a way to throw exceptions with more information?